BaseLib/FileTools.cpp

```diff
@@ -39,6 +39,107 @@ bool IsFileExisting(const std::string &strFilename)
+        auto const pos_curly_brace_close =
+            in.find_first_of(close_char, pos_curly_brace_open);
+        if (pos_curly_brace_close != std::string::npos)
+        {
+            return std::make_tuple(
+                in.substr(pos_curly_brace_open + 1,
+                          pos_curly_brace_close - (pos_curly_brace_open + 1)),
+                pos_curly_brace_open, pos_curly_brace_close);
+        }
+    }
+    return std::make_tuple("", std::string::npos, std::string::npos);
+}
+
+std::pair<bool, std::string> containsKeyword(std::string const& str,
+                                             std::string const& keyword)
+{
+    auto const position = str.find(keyword);
+    if (position != std::string::npos)
+    {
+        return std::make_pair(true, str.substr(0, position));
+    }
+    return std::make_pair(false, "");
+}
+
+template <typename T>
+bool substituteKeyword(std::string& result, std::string& parenthesized_string,
+                       std::string::size_type begin, std::string::size_type end,
+                       std::string const& keyword, T& data)
+{
+    std::unordered_map<std::type_index, std::string> type_specification;
+    type_specification[std::type_index(typeid(int))] = "d";
+    type_specification[std::type_index(typeid(double))] = "f";
+
+    bool contains_keyword = false;
+    std::string precision_specification = "";
+    std::tie(contains_keyword, precision_specification) =
+        containsKeyword(parenthesized_string, keyword);
+
+    if (contains_keyword)
+    {
+        std::string const generated_fmt_string =
+            "{" + precision_specification +
+            type_specification[std::type_index(typeid(data))] + "}";
+        result = result.substr(0, begin) +
+                 fmt::format(generated_fmt_string, data) +
+                 result.substr(end + 1, result.length() - (end + 1));
+        return true;
+    }
+    return false;
+}
+
+std::string constructFormattedFileName(std::string const& format_specification,
+                                       int const process_id,
+                                       int const timestep,
+                                       double const t)
+{
+    char const open_char = '{';
+    char const close_char = '}';
+    std::string::size_type begin = 0;
+    std::string::size_type end = std::string::npos;
+    std::string result = format_specification;
+
+    while (begin != std::string::npos)
+    {
+        auto length_before_substitution = result.length();
+        // find next parenthesized string
+        std::string str = "";
+        std::tie(str, begin, end) =
+            getParenthesizedString(result, open_char, close_char, begin);
+        if (!substituteKeyword(result, str, begin, end, "timestep", timestep))
+        {
+            if (!substituteKeyword(result, str, begin, end, "time", t))
+            {
+                substituteKeyword(result, str, begin, end, "process",
+                                  process_id);
```

Comment:
```suggestion
                substituteKeyword(result, str, begin, end, "processId",
                                  process_id);
```
or maybe `process_id` would be better. Under `process` I guessed a name of the process rather than its id.

Reply:
Sounds reasonable, will change this.